src/task1/CubeWindow.cpp

```diff
@@ -0,0 +1,139 @@
+            6, 6, 4, 0, 1, 1,
+            2, 2, 3, 7, 5
+    };
+
+
+} // namespace
+
+namespace fgl {
+
+    void CubeWindow::init() {
+        program_ = std::make_unique<QOpenGLShaderProgram>(this);
+        program_->addShaderFromSourceFile(QOpenGLShader::Vertex,
+                                          ":/Shaders/square.vs");
+        program_->addShaderFromSourceFile(QOpenGLShader::Fragment,
+                                          ":/Shaders/square.fs");
+        program_->link();
+
+        posAttr_ = program_->attributeLocation("posAttr");
+        assert(posAttr_ != -1);
+        colAttr_ = program_->attributeLocation("colAttr");
+        assert(colAttr_ != -1);
+        matrixUniform_ = program_->uniformLocation("matrix");
+        assert(matrixUniform_ != -1);
+
+        glEnable(GL_DEPTH_TEST);
+        glEnable(GL_CULL_FACE);
+
+    }
+
+    void CubeWindow::render() {
+        const auto retinaScale = devicePixelRatio();
+        glViewport(0, 0, width() * retinaScale, height() * retinaScale);
+
+
+        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+        program_->bind();
+
+
+        QMatrix4x4 matrix;
+        matrix.perspective(60.0f, 4.0f / 3.0f, 0.1f, 100.0f);
+        matrix.translate(0, 0, -2);
+        matrix.rotate(100.0 * frame_ / screen()->refreshRate(), rotationAxis);
+
+        program_->setUniformValue(matrixUniform_, matrix);
+        program_->setUniformValue("col", square_color );
+        unsigned int vbo;
+        glGenBuffers(1, &vbo);
+        glBindBuffer(GL_ARRAY_BUFFER, vbo);
+        glBufferData(GL_ARRAY_BUFFER, 8 * sizeof(VertexData), gvertices, GL_STATIC_DRAW);
+
+
+        glVertexAttribPointer(posAttr_, 3, GL_FLOAT, GL_FALSE, sizeof(VertexData), 0);
+        glVertexAttribPointer(colAttr_, 3, GL_FLOAT, GL_FALSE, sizeof(VertexData), reinterpret_cast<char *>(sizeof(QVector3D) ));
+
+        glEnableVertexAttribArray(posAttr_);
+        glEnableVertexAttribArray(colAttr_);
+
+        unsigned int v_ibo;
+        glGenBuffers(1, &v_ibo);
+        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, v_ibo);
+        glBufferData(GL_ELEMENT_ARRAY_BUFFER, 34 * sizeof(unsigned int), v_indicies.data(), GL_STATIC_DRAW);
+
+        glDrawElements(GL_TRIANGLE_STRIP, 34, GL_UNSIGNED_INT, nullptr);
+        glDisableVertexAttribArray(colAttr_);
+        glDisableVertexAttribArray(posAttr_);
+
+        program_->release();
+
+        ++frame_;
+    }
+
+    void CubeWindow::keyPressEvent(QKeyEvent *event){
+        if (event->key() == Qt:: Key_Tab){
+            QColor color = QColorDialog::getColor();
+            square_color = QVector4D(color.red() / 255.0, color.green()/255.0, color.blue()/255.0, 1);
+        }
+    }
+
+    void CubeWindow::mousePressEvent(QMouseEvent *e)
+    {
+
+        mousePressPosition = QVector2D(e->localPos());
+    }
+
+    void CubeWindow::mouseReleaseEvent(QMouseEvent *e)
+    {
+
+        QVector2D diff = QVector2D(e->localPos()) - mousePressPosition;

```

Comment:
```suggestion
        const autodiff = QVector2D(e->localPos()) - mousePressPosition;

```
и ниже тоже

https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/